### DISCUSSION TITLE
When receiving a RTP stream, will pion correct misordered packets?

### DISCUSSION THREAD

**Discussion by birla** [pion/webrtc#2249] [Q&A] [answered]
I'm trying to use this library for building a RTP Server to recieve multiple RTP streams. I want to ensure that out of order packets are reordered before they are consumed, does pion handle this internally or could anyone point me to a sample?

Much appreciated, thanks

**Comment by alexpokotilo** [comment, marked as answer]
@birla 
what do you find during your tests ? As far as I see ReadRTP will give you packets in order they received and hence sorting reorderings(if you really need this) is your responsibility. If I am wrong, would be cool to be corrected

**Comment by Sean-Der** [comment]
Hi @birla 

@alexpokotilo is correct :) By default Pion doesn't delay or de-jitter packets.

Pion does provide the [SampleBuilder](https://github.com/pion/webrtc/tree/master/pkg/media/samplebuilder). You can run your RTP packets through it and it will re-order them + delay a configured amount of time for late packets.

**Reply by alexpokotilo** [reply]
Hi Sean,
thanks for confirmation! I see SampleBuilder works for re-sent packages after NACK too, right ?
I wish I have this answer in August as I made my own pure implementation. I will switch to SampleBuilder in anyway

**Reply by birla** [reply]
Thanks Alex and Sean!

**Reply by alexpokotilo** [reply]
> Hi @birla
> 
> @alexpokotilo is correct :) By default Pion doesn't delay or de-jitter packets.
> 
> Pion does provide the [SampleBuilder](https://github.com/pion/webrtc/tree/master/pkg/media/samplebuilder). You can run your RTP packets through it and it will re-order them + delay a configured amount of time for late packets.

I was able to use SampleBuilder to support re-order and packet loss for audio and all supported video codecs.
Thank you @Sean-Der

**Comment by half2me** [comment]
Is there a way to use the sample builder together with `oggwriter`?

Atm I'm just initializing the oggwriter and passing the RTP packets into it like this:
This gives me a good recording in ideal circumstances, but I'm worried about cases where RTP packets could be out of order.
```go
writer, _ := oggwriter.New("recording.ogg", 48000, 2)

buf := make([]byte, 1600)
for {
  n, _, _ := conn.ReadFrom(buf)
  rtpPacket := &rtp.Packet{}
  _ = rtpPacket.Unmarshal(buf[:n])
  _ = m.writer.WriteRTP(rtpPacket)
}

writer.Close()
```

If I use the samplebuilder:
```go
builder := samplebuilder.New(10, &codecs.OpusPacket{}, 48000) // how do I set the num of channels?

buf := make([]byte, 1600)
for {
  n, _, _ := conn.ReadFrom(buf)
  rtpPacket := &rtp.Packet{}
  _ = rtpPacket.Unmarshal(buf[:n])
  builder.Push(rtpPacket)

  sample := builder.Pop()
  if sample != nil {
    // how to I write this sample to a file?
  }
}

// how do I flush the last remaining packets out?
```

**Reply by shushenghong** [reply]
@half2me i have the same question, after search the code 
inside the samplebuilder / oggwriter
https://github.com/pion/webrtc/blob/1c4bc798fe759e77a98589655f29e353de4698df/pkg/media/samplebuilder/samplebuilder.go#L285
https://github.com/pion/webrtc/blob/1c4bc798fe759e77a98589655f29e353de4698df/pkg/media/oggwriter/oggwriter.go#L194
they also use `opusPacket.Unmarshal`, they almost the same, we need write a method to mock`createPage` with the unmarshaled data
